packages/contracts/contracts/L1/L2OutputOracle.sol

```diff
@@ -0,0 +1,113 @@
+        startingBlockTimestamp = block.timestamp; // solhint-disable not-rely-on-time
+
+        _transferOwnership(sequencer);
+    }
+
+    /**
+     * Accepts an L2 output checkpoint and the timestamp of the corresponding L2
+     * block. The timestamp must be equal to the current value returned by
+     * `nextTimestamp()` in order to be accepted.
+     * @param _l2Output The L2 output of the checkpoint block.
+     * @param _l2timestamp The L2 block timestamp that resulted in _l2Output.
+     * @param _blockhash A block hash which must be included in the current chain.
+     * @param _blocknumber The block number with the specified block hash.
+     */
+    function appendL2Output(bytes32 _l2Output, uint256 _l2timestamp, bytes32 _blockhash, uint256 _blocknumber) external payable onlyOwner {
+        require(_l2timestamp < block.timestamp, "Cannot append L2 output in future");
+        require(_l2timestamp == nextTimestamp(), "Timestamp not equal to next expected timestamp");
+        require(_l2Output != bytes32(0), "Cannot submit empty L2 output");
+
+
+        if (_blockhash != bytes32(0)) {
+            // This check allows the sequencer to append an output based on a given L1 block,
+            // without fear that it will be reorged out.
+            // It will also revert if the blockheight provided is more than 256 blocks behind the
+            // chain tip (as the hash will return as zero). This does open the door to a grieffing
+            // attack in which the sequencer's submission is censored until the block is no longer
+            // retrievable, if the sequencer is experiencing this attack it can simply leave out the
+            // blockhash value, and delay submission until it is confident that the L1 block is
+            // finalized.
+            require(
+                blockhash(_blocknumber) == _blockhash,
+                "Blockhash does not match the hash at the expected height."
+            );
+        }
+
+        l2Outputs[_l2timestamp] = _l2Output;
+        latestBlockTimestamp = _l2timestamp;
+
+        emit l2OutputAppended(_l2Output, _l2timestamp);
+    }
+
+    /**
+     * Computes the timestamp of the next L2 block that needs to be
+     * checkpointed.
+     */
+    function nextTimestamp() public view returns (uint256) {
+        return latestBlockTimestamp + submissionInterval;
+    }
+
+    /**
+     * Computes the L2 block number given a target L2 block timestamp.
+     * @param _l2timestamp The L2 block timestamp of the target block.
+     */
+    function computeL2BlockNumber(uint256 _l2timestamp) external view returns (uint256) {
+        require(
+            _l2timestamp >= startingBlockTimestamp,
+            "Timestamp prior to startingBlockTimestamp"
+        );
+        // If _l2timestamp == startingBlockTimestamp, then the L2BlockNumber should be
+        // historicalTotalBlocks + 1
+        return historicalTotalBlocks + 1 + (_l2timestamp - startingBlockTimestamp) / l2BlockTime;
```

Comment:
The previous mock had what I think was an off-by-one error in it, but I'd appreciate if someone can sanity check that this comment is the correct behavior.

Reply:
I would like to see some fuzz tests for this or something to see if there are any problems with precision